api/Registry: Add component module detection

javajawa opened the pull request:
Modules which provide the `mewbot-v1` entrypoint in their module declaration can now be automatically detected and loaded. This will allow, in the future, for systems to retrieve a list of components that are available for use.

Additionally, the documentation for the Registry has been improved

javajawa committed:
api/Registry: Add component module detection

Modules which provide the mewbot-v1 entrypoint in their module declaration
can now be automatically detected and loaded. This will allow, in the future,
for systems to retrieve a list of components that are available for use.

Additionally, the documentation and tests for the Registry has been improved.

javajawa commented:
Those were meant to be one commit; will squash when I add the relevant tests.

ajCameron commented on src/mewbot/api/registry.py:
```diff
@@ -121,14 +165,48 @@ def do_register(api: Type[Component]) -> Type[Component]:
+
+        for val in mcs._detect_api_versions(component):
             return val
 
```
Suggested change - as far as I can tell, for well designed code which is behaving how we'd prefer, this should never trip more than once (if nothing else, there's a check elsewhere which makes it unlikely).

Thus I propose it should error, instead of returning the first matched value, if there is more than one match. Seems less likely to produce unexpected behavior and a safer behavior. People could still work around it, but at least they would know they where doing something a bit didgy. 

(But there might well be some use case I'm missing which would be a shame to loose, so I'm not married to the idea).

javajawa replied:
I've written up a change (754854a) that makes this examine the list, but I'm wondering what situations a second iteration could happen excluding people fiddling with subclass hooks -- and I'm no longer sure its worth defending against that?

ajCameron replied:
Hmm. Yeah - it _should_ never occur unless, as you say, someone has messed with subclass hooks.

It just seems like the sort of assumption that might trip someone up at some point (it should match once and only once, but, currently, it silently works if, afact, there's more than one match).

It's probably not an issue! I'd only suggest closing it off out of an abundance of caution.

exachixkitsune replied:
My opinion would be if it's a potential error state - even if we don't know how the error could occur - it should report an error; but then that's just in the effort of robustness not efficiency.

javajawa replied:
754854a has been included in the tree

ajCameron commented:
(Overall comment - code looks really good! Very elegant)

javajawa committed:
api/Registry: Ensure only one API version

This causes the ComponentRegistry.api_version to raise a ValueError
if the Component type doesn't have exactly one API version

exachixkitsune commented on src/mewbot/api/registry.py:
```diff
@@ -121,14 +165,48 @@ def do_register(api: Type[Component]) -> Type[Component]:
+    @staticmethod
+    def load_and_register_modules(name: Optional[str] = None) -> Iterable[Any]:
+        """
+        Load modules from setuptools that declare a `mewbot-v1` entrypoint.
```
Does this comment want to say "mewbot-v1" or "setuptools that declare an entrypoint matching the current api"
```suggestion
        Load modules from setuptools that declare a entrypoint with a valid API.
```

javajawa replied:
Hmm, interesting. If we're going for that, I'd probably say `supported`?

exachixkitsune replied:
Sounds reasonable to me
```suggestion
        Load modules from setuptools that declare an entrypoint with a supported API.
```

exachixkitsune commented on src/mewbot/api/registry.py:
```diff
@@ -121,14 +165,48 @@ def do_register(api: Type[Component]) -> Type[Component]:
+        for distribution in importlib_metadata.distributions():  # type: ignore
+            for entry_point in distribution.entry_points:
+                if entry_point.group not in API_DISTRIBUTIONS:
+                    continue
```
On the thought of error dectection, is it worth saying "This item's not being picked up because of this reason"? so we're not like "Why is my module not loading? :("

javajawa replied:
That message would occur for every python module installed in the current context, which I don't think would be productive for anyone?

exachixkitsune replied:
ah. that's fair yeah maybe let's not do that. Is there some other confirmation of what modules load?

javajawa replied:
I'm not quite sure what you mean by "confirmation"? This isn't intended as an interactive process, and even having a name filter is mostly for having the same signature as `pluggy`

exachixkitsune replied:
Resolved via direct discussion

exachixkitsune reviewed (dismissed):
I think I approve of a bunch of this.

javajawa committed:
api/Registry: Split plugin loading functions

Loading all appearant plugins, and loading a specific plugin by name,
have been split into two different functions with different expectations.

The new `require_package` function will throw an error if the plugin
is not available, doesn't not present a compatible API, or fails to
load during import.

The ability to filter by name has been removed from
`load_and_register_modules`, which will now only load all packages
it is able to load based on the entry point information.

exachixkitsune reviewed (approved):
Seems ok to me!
